Support returning hidden state in CLIPTextEncoder

22quinn opened the pull request:
Summary:
Added a `return_hidden_state` arg in CLIPTextEncoder. If set to True, forward will return a tuple of final embedding and the last hidden state.

Test plan:
pytest tests/models/clip/test_text_encoder.py

cc @abhinavarora Moved code to here for easier review and testing.

ankitade commented on torchmultimodal/models/clip/text_encoder.py:
```diff
@@ -120,11 +127,13 @@ def forward(self, text: Tensor) -> Tensor:
         # embeddings now has size [bs, embedding_dim]
 
+        if self.return_hidden_state:
+            return embeddings, hidden_state
```
nit: use named tuple

ankitade commented on tests/models/clip/test_text_encoder.py:
```diff
@@ -117,6 +121,47 @@ def test_forward(self, start):
+        assert isinstance(text_encoder, torch.nn.Module)
+
+        actual_clip_init, actual_hidden_state = text_encoder(text)
+        print(actual_hidden_state)
```
remove

ankitade commented on tests/models/clip/test_text_encoder.py:
```diff
@@ -117,6 +121,47 @@ def test_forward(self, start):
+        expected_hidden_state = torch.Tensor(
+            [
+                [
+                    [6.348165e-01, -4.137459e-02, -1.604239e00, 1.010798e00],
```
nit: can we limit to 4 decimal places given atol is being set

ankitade commented on torchmultimodal/models/clip/text_encoder.py:
```diff
@@ -77,6 +82,8 @@ def __init__(
         if use_clip_init:
             self.initialize_parameters()
 
+        self.return_hidden_state = return_hidden_state
```
you can make this an arg to fwd

abhinavarora commented on torchmultimodal/models/clip/text_encoder.py:
```diff
@@ -120,11 +127,13 @@ def forward(self, text: Tensor) -> Tensor:
-        embeddings = self.ln_final(embeddings)
+        hidden_state = self.ln_final(embeddings)
         # take features from the eot embedding (the highest number in each sequence)
         embeddings = self.projection(
```
I would suggest renaming this to something like `projected_embeddings` to avoid re-using the `embeddings` name as it was used before.

22quinn commented:
Thanks both for the review! I've addressed all comments. buck test looks good.

> buck test @mode/dev-nosan //torchmultimodal/tests
Tests finished: Pass 317. Fail 0. Fatal 0. Skip 6. Build failure 0

ankitade commented:
can you import the diff and land it there
